### ISSUE TITLE
Some regex_replace like function to produce dynamic labels

### ISSUE THREAD

**Issue by mouhong** [google/mtail#291]
I would like to extract url **route pattern** as labels. Say, we have a http request: 

```
GET /v1/read/10001
```

The 10001 is the entity ID, so the **route pattern** is `v1/read/:num`. It's not recommended to use high cardinality labels in Prometheus, and we often want to see how some specific API performs, that's why I would like to use route pattern as labels instead of request uri.

Question: Is it possible to do this in a generic way? I mean, when new urls are added, we don't have to modify mtail programs. Currently I have to do this in a hard-coded way:

```
hidden text route
counter http_requests_total by method, status, route

const ROUTE_READ_ARTICLE / \/v\d\/read\/\d+/
const ROUTE_POST_READLOGS /POST \/v\d\/my\/readlogs/

// + ROUTE_READ_ARTICLE {
    route = "read"
}
// + ROUTE_POST_READLOGS {
    route = "post_readlogs"
}
otherwise {
    route = "unknown"
}

/\[(?P<date>[^\]]+)\]/ {
    strptime($date, "02/Jan/2006:15:04:05 -0700")

    /\] "(?P<method>\S+) .+ HTTP\/\S+ (?P<status>\d+) / {
        http_requests_total[$method, $status, route]++
    }
}
```

The expected way (pseudocode):

```
hidden text route
counter http_requests_total by method, status, route

/\[(?P<date>[^\]]+)\]/ {
    strptime($date, "02/Jan/2006:15:04:05 -0700")

    /\] "(?P<method>\S+) (?P<url>\S+) .+ HTTP\/\S+ (?P<status>\d+) / {
        # here I replace the integers in the url path to product the route pattern
        route = regex_replace($url, /\/\d+\//, "/:num/")
        inovel_http_requests_total[$method, $status, route]++
    }
}
``` 

In the above pseudocode, I use the `regex_replace` function to transform urls like `v1/read/10001` to `v1/read/:num`. Note that this time the mtail program is **no longer hard-coded**. It can still work for new urls like `v1/users/10002/orders` (route pattern: `v1/users/:num/orders`).

**Comment by jaqx0r** [comment]
Do you need to include ":num" in the route to disambiguate it from other
routes?

Could you instead use just v1/read instead of v1/read/:num, and
v1/read/orders instead of v1/read/:num/orders?

Because you could do it with your second way:

`
counter http_requests_total by method, status, route

/\[(?P<date>[^\]]+)\]/ {
    strptime($date, "02/Jan/2006:15:04:05 -0700")

    /\] "(?P<method>\S+) (?P<route>[^\d]+)(\d+(?P<opt_suffix>\S+))? .+
HTTP\/\S+ (?P<status>\d+) / {
        inovel_http_requests_total[$method, $status,
$route+":num"+$opt_suffix]++
    }
}
`

That'd need a bit of work to make sure the route group's pattern is
correct, but it's basically the same as accepting everything that's not a
\d.  It'll trip over your v1 though, so like I said, needs some work.

What happens if the route has a second number or identifier in the URL?

**Comment by mouhong** [comment]
@jaqx0r The basic idea is that, I would like to replace all integer segments in the url into the placeholder ":num". So `v1/read/10001` becomes `v1/read/:num`, and `v1/users/1001/orders/2001` becomes `v1/users/:num/orders/:num`. 

I would like to display top K slowest "routes" in our Grafana dashboard. `v1/read/:num` might be the slowest route, but `v1/read/10001` might not, that's why I need the **route pattern** instead of the actual request url.

We applied RESTful pattern in our API design, and our entity ids are mostly integers, replacing integers to placeholder ":num" is the easiest way to extract url route pattern from urls. I don't want to hard-code our 100+ routes one by one :(.

Others might use UUIDs instead of integers in the url, but it's still easy to extract route patterns from url, as long as there's some "pattern" in the urls and we have the ability to **replace** matched groups into placeholders.

In short,  I think I need some `sed` like function.

**Comment by jaqx0r** [comment]
I've just added [`subst`](https://google.github.io/mtail/Programming-Guide.html#parsing-numbers-with-extra-characters) to the language which does string substitution.  I still don't understand why the suggestion I made doesn't suffice here, because it's identical to a regex_replace in mechanism -- a regular expression match followed by a substitution.  But a seeing as how `subst` was easy, I'm going to attempt to add support for supplying a regular expression pattern as the first parameter.
